src/Renci.SshNet/SftpClient.cs

```diff
@@ -2074,60 +2074,68 @@ private IEnumerable<FileInfo> InternalSynchronizeDirectories(string sourcePath,
 
             var sourceDirectory = new DirectoryInfo(sourcePath);
 
-            var sourceFiles = FileSystemAbstraction.EnumerateFiles(sourceDirectory, searchPattern).ToList();
-            if (sourceFiles.Count == 0)
-                return uploadedFiles;
-
-            #region Existing Files at The Destination
-
-            var destFiles = InternalListDirectory(destinationPath, null);
-            var destDict = new Dictionary<string, ISftpFile>();
-            foreach (var destFile in destFiles)
+            using (var sourceFiles = sourceDirectory.EnumerateFiles(searchPattern).GetEnumerator())
```

Comment:
I had a closer look at the implementation now.
We do not use a `foreach` to iterate over the result returned by `sourceDirectory.EnumerateFiles(searchPattern)` to avoid initializing the `destFiles` dictionary if there are no local files matching the pattern.